pyteal/compiler/subroutines.py

```diff
@@ -108,63 +112,89 @@ def spillLocalSlotsDuringRecursion(
                 # restore the local slots after returning from the subroutine. This prevents a
                 # reentry into the current subroutine from modifying variables we are currently
                 # using.
+
+                digArgs = True
+                coverSpilledSlots = False
+                uncoverArgs = False
+                if coverAvailable:
+                    digArgs = False
+                    if len(slots) < numArgs:
+                        coverSpilledSlots = True
+                    else:
+                        uncoverArgs = True
```

Comment:
Could you please explain what this condition checks? I am not sure in what scenario `len(slots) < numArgs` is triggered, since the current check only cover condition `len(slots) == numArgs`.

Sorry in advance for the questions...

Reply:
That's a good question. Before a recursive subroutine call, the arguments for the call are already on the stack. The compilers need to spill all local slots to the stack, somehow pull the arguments to the top of the stack, and then invoke the subroutine (and afterwards restore the slots).

There are 2 options to "somehow pull the arguments to the top of the stack":
1. Use `cover` to place each spilled slot below the arguments. E.g. with 3 args and 2 local slots:
```
arg 1
arg 2
arg 3
# the stack is now [arg1, arg2, arg3]

load 0
cover 3
load 1
cover 3
# the stack is now [slot0, slot1, arg1, arg2, arg3]
```
(I just realized the argument should be `numArgs`, not `len(slots) + numArgs - 1` 😅 -- thank you for questioning the lack of tests!)

2. Use `uncover` to pull up each argument above the spilled slots. E.g. with 3 args and 2 local slots:
```
arg 1
arg 2
arg 3
# the stack is now [arg1, arg2, arg3]

load 0
load 1
# the stack is now [arg1, arg2, arg3, slot0, slot1]

uncover 4
uncover 4
uncover 4
# the stack is now [slot0, slot1, arg1, arg2, arg3]
```

They are both equivalent, but depending on the number of arguments vs spilled slots, one might use more opcodes than the other -- in the examples using `cover` is more efficient.

I will add more tests to make sure all cases are evaluated.

Reply:
All good all good, I believe once this is resolved, it should be good to merge. Thanks for adding more tests!

Reply:
More tests added, but the implementation was actually correct. I thought it used the wrong value, but it uses `numArgs`

Reply:
I see, the point here is to pick the method using least operation. Very neat.